skopt/optimizer/base.py

```diff
@@ -186,129 +172,81 @@ def base_minimize(func, dimensions, base_estimator,
     n_init_func_calls = len(x0) if y0 is None else 0
     n_total_init_calls = n_random_starts + n_init_func_calls
 
-    if n_calls <= 0:
-        raise ValueError("Expected `n_calls` > 0, got %d" % n_calls)
-
-    if n_random_starts < 0:
+    if n_calls < n_total_init_calls:
         raise ValueError(
-            "Expected `n_random_starts` >= 0, got %d" % n_random_starts)
+            "Expected `n_calls` >= %d, got %d" % (n_total_init_calls,
+                                                  n_calls))
 
     if n_random_starts == 0 and not x0:
-        raise ValueError("Either set `n_random_starts` > 0, or provide `x0`")
-
-    if n_calls < n_total_init_calls:
-        raise ValueError(
-            "Expected `n_calls` >= %d, got %d" % (n_total_init_calls, n_calls))
+        raise ValueError("Either set `n_random_starts` > 0,"
+                         " or provide `x0`")
 
     callbacks = check_callback(callback)
     if verbose:
         callbacks.append(VerboseCallback(
             n_init=n_init_func_calls, n_random=n_random_starts,
             n_total=n_calls))
 
-    if y0 is None and x0:
-        y0 = []
+    # User suggested points at which to evaluate the objective first
+    if x0 and y0 is None:
         for i, x in enumerate(x0):
-            y0.append(func(x))
-            curr_res = create_result(x0[:i + 1], y0, space, rng, specs)
+            fit_model = i == len(x0) - 1
+            y = func(x)
+            optimizer.tell(x, y, fit_model=fit_model)
 
             if callbacks:
+                result = create_result(optimizer.Xi, optimizer.yi,
+                                       optimizer.space, optimizer.rng,
+                                       specs, optimizer.models)
                 for c in callbacks:
-                    c(curr_res)
+                    c(result)
+
+    # User is god and knows the value of the objective at these points
+    elif x0 and y0 is not None:
+        if not (isinstance(y0, Iterable) or isinstance(y0, numbers.Number)):
+            raise ValueError(
+                "`y0` should be an iterable or a scalar, got %s" % type(y0))
 
-    elif x0:
         if isinstance(y0, Iterable):
             y0 = list(y0)
         elif isinstance(y0, numbers.Number):
             y0 = [y0]
-        else:
-            raise ValueError(
-                "`y0` should be an iterable or a scalar, got %s" % type(y0))
+
         if len(x0) != len(y0):
             raise ValueError("`x0` and `y0` should have the same length")
+
         if not all(map(np.isscalar, y0)):
             raise ValueError(
                 "`y0` elements should be scalars")
-    else:
-        y0 = []
 
-    # Random function evaluations.
-    X_rand = space.rvs(n_samples=n_random_starts, random_state=rng)
-    Xi = x0 + X_rand
-    yi = y0
+        for i, (x, y) in enumerate(zip(x0, y0)):
+            fit_model = i == len(x0) - 1
+            optimizer.tell(x, y, fit_model=fit_model)
```

Comment:
if you provide `x0`, `y0` and `n_random_starts` we add the observations you provide, and then do an additional `n_random_starts`. The case where just `x0` is provided (but not `y0` is handled on line 193.